File: xposed-pine/src/org/pixel/customparts/hooks/systemui/DozeTapDozeHook.java
package org.pixel.customparts.hooks.systemui;

import android.content.Context;
import android.os.PowerManager;
import android.os.SystemClock;
import de.robv.android.xposed.XC_MethodHook;
import de.robv.android.xposed.XposedBridge;
import de.robv.android.xposed.XposedHelpers;
import org.pixel.customparts.core.BaseHook;

import java.lang.reflect.Method;
import java.util.Collection;

public class DozeTapDozeHook extends BaseHook {
    
    private static final int WAKE_REASON_TAP = 15;

    private int dozeTapReason = 9; // Default fallback

    @Override
    public String getHookId() {
        return "DozeTapDozeHook";
    }

    @Override
    public int getPriority() {
        return 70;
    }

    @Override
    public boolean isEnabled(Context context) {
        return true;
    }

    @Override
    protected void onInit(ClassLoader classLoader) {
        resolveTapReason(classLoader);
        hookDozeTriggers(classLoader);
    }

    private void resolveTapReason(ClassLoader classLoader) {
        try {
            Class<?> dozeLogClass = XposedHelpers.findClass("com.android.systemui.doze.DozeLog", classLoader);
            dozeTapReason = XposedHelpers.getStaticIntField(dozeLogClass, "REASON_SENSOR_TAP");
        } catch (Throwable t) {
            log("DozeTapDozeHook: Using default REASON_SENSOR_TAP = " + dozeTapReason);
        }
    }

    private void hookDozeTriggers(ClassLoader classLoader) {
        try {
            Class<?> dozeTriggersClass = XposedHelpers.findClass(
                "com.android.systemui.doze.DozeTriggers",
                classLoader
            );

            XposedBridge.hookAllMethods(dozeTriggersClass, "onSensor", new XC_MethodHook() {
                @Override
                protected void beforeHookedMethod(MethodHookParam param) {
                    try {
                        // Ищем первый аргумент типа Int (pulseReason)
                        int pulseReason = -1;
                        for (Object arg : param.args) {
                            if (arg instanceof Integer) {
                                pulseReason = (Integer) arg;
                                break;
                            }
                        }

                        if (pulseReason != dozeTapReason) return;

                        Object dozeTriggers = param.thisObject;
                        Context context = (Context) XposedHelpers.getObjectField(dozeTriggers, "mContext");
                        if (context == null) return;
                        
                        if (!isSettingEnabled(context, DozeTapManager.KEY_HOOK)) return;

                        // Ищем float аргументы (x, y)
                        float screenX = -1f;
                        float screenY = -1f;
                        int floatCount = 0;
                        
                        for (Object arg : param.args) {
                            if (arg instanceof Float) {
                                if (floatCount == 0) screenX = (Float) arg;
                                else if (floatCount == 1) screenY = (Float) arg;
                                floatCount++;
                            }
                        }

                        if (floatCount < 2) return;

                        int timeout = getIntSetting(context, DozeTapManager.KEY_TIMEOUT, DozeTapManager.DEFAULT_TIMEOUT);
                        final int tapReason = pulseReason;
                        final float tapX = screenX;
                        final float tapY = screenY;
                        
                        DozeTapManager.TapResult tapResult = DozeTapManager.processTap(
                            context,
                            tapX,
                            tapY,
                            true,
                            timeout,
                            new Runnable() {
                                @Override
                                public void run() {
                                    prepareTapSensorForNextTap(dozeTriggers);
                                }
                            },
                            new Runnable() {
                                @Override
                                public void run() {
                                    wakeFromSystemDoze(dozeTriggers, tapReason, tapX, tapY);
                                }
                            }
                        );

                        if (tapResult != DozeTapManager.TapResult.IGNORED) {
                            param.setResult(null); // Блокируем выполнение оригинального метода
                        }
                    } catch (Throwable t) {
                        logError("DozeTapDozeHook: Error in onSensor", t);
                    }
                }
            });

            log("DozeTapDozeHook: Hook applied successfully");
        } catch (Throwable e) {
            logError("DozeTapDozeHook: Failed to apply hook", e);
        }
    }

    private void prepareTapSensorForNextTap(Object dozeTriggers) {
        enableNativeTapSensorReregister(dozeTriggers);
        reregisterTapSensor(dozeTriggers);
    }

    private boolean enableNativeTapSensorReregister(Object dozeTriggers) {
        try {
            Object dozeSensors = XposedHelpers.getObjectField(dozeTriggers, "mDozeSensors");
            if (dozeSensors == null) return false;

            Object triggerSensors = XposedHelpers.getObjectField(dozeSensors, "mTriggerSensors");
            if (triggerSensors == null) return false;

            boolean updated = false;
            if (triggerSensors instanceof Object[]) {
                for (Object sensor : (Object[]) triggerSensors) {
                    updated |= enableNativeTapSensorReregisterForSensor(sensor);
                }
            } else if (triggerSensors instanceof Collection) {
                for (Object sensor : (Collection<?>) triggerSensors) {
                    updated |= enableNativeTapSensorReregisterForSensor(sensor);
                }
            }
            return updated;
        } catch (Throwable t) {
            logError("DozeTapDozeHook: enableNativeTapSensorReregister failed", t);
            return false;
        }
    }

    private boolean enableNativeTapSensorReregisterForSensor(Object sensor) {
        if (sensor == null) return false;
        try {
            int reason = XposedHelpers.getIntField(sensor, "mPulseReason");
            if (reason != dozeTapReason) return false;
            XposedHelpers.setBooleanField(sensor, "mImmediatelyReRegister", true);
            log("DozeTapDozeHook: native tap sensor re-register enabled");
            return true;
        } catch (Throwable t) {
            log("DozeTapDozeHook: native re-register enable failed: " + t.getMessage());
            return false;
        }
    }

    private void wakeFromSystemDoze(Object dozeTriggers, int reason, float screenX, float screenY) {
        try {
            Object host = XposedHelpers.getObjectField(dozeTriggers, "mDozeHost");
            if (host != null) {
                XposedHelpers.callMethod(host, "onSlpiTap", screenX, screenY);
            }
        } catch (Throwable ignored) {
        }

        try {
            Object machine = XposedHelpers.getObjectField(dozeTriggers, "mMachine");
            if (machine != null) {
                XposedHelpers.callMethod(machine, "wakeUp", reason);
                log("DozeTapDozeHook: double tap woke via DozeMachine");
                return;
            }
        } catch (Throwable t) {
            log("DozeTapDozeHook: DozeMachine wake failed: " + t.getMessage());
        }

        try {
            Context context = (Context) XposedHelpers.getObjectField(dozeTriggers, "mContext");
            if (context == null) return;
            PowerManager powerManager = (PowerManager) context.getSystemService(Context.POWER_SERVICE);
            if (powerManager != null) {
                Method wakeUp = PowerManager.class.getMethod(
                    "wakeUp", long.class, int.class, String.class);
                wakeUp.invoke(powerManager, SystemClock.uptimeMillis(),
                    WAKE_REASON_TAP, "PixelPartsDozeDoubleTap");
                log("DozeTapDozeHook: double tap woke via PowerManager fallback");
            }
        } catch (Throwable t) {
            logError("DozeTapDozeHook: PowerManager wake fallback failed", t);
        }
    }

    private void reregisterTapSensor(Object dozeTriggers) {
        try {
            Object dozeSensors = XposedHelpers.getObjectField(dozeTriggers, "mDozeSensors");
            if (dozeSensors == null) return;
            
            Object triggerSensors = XposedHelpers.getObjectField(dozeSensors, "mTriggerSensors");
            if (triggerSensors == null) return;

            if (triggerSensors instanceof Object[]) {
                for (Object sensor : (Object[]) triggerSensors) {
                    if (sensor != null) checkAndResetSensor(sensor);
                }
            } else if (triggerSensors instanceof Collection) {
                for (Object sensor : (Collection<?>) triggerSensors) {
                    if (sensor != null) checkAndResetSensor(sensor);
                }
            }
        } catch (Throwable t) {
             logError("DozeTapDozeHook: reregisterTapSensor failed", t);
        }
    }

    private void checkAndResetSensor(Object sensor) {
        try {
            int reason = XposedHelpers.getIntField(sensor, "mPulseReason");
            if (reason == dozeTapReason) {
                Method method = sensor.getClass().getDeclaredMethod("setListening", boolean.class);
                method.setAccessible(true);
                method.invoke(sensor, false);
                method.invoke(sensor, true);
                log("DozeTapDozeHook: Sensor re-registered OK");
            }
        } catch (Throwable e) {
            log("DozeTapDozeHook: checkAndResetSensor failed: " + e.getMessage());
            try {
                // Fallback method
                XposedHelpers.setBooleanField(sensor, "mRequested", false);
                XposedHelpers.callMethod(sensor, "updateListening");
                XposedHelpers.setBooleanField(sensor, "mRequested", true);
                XposedHelpers.callMethod(sensor, "updateListening");
                log("DozeTapDozeHook: Sensor re-registered via fallback OK");
            } catch (Throwable e2) {
                log("DozeTapDozeHook: Fallback re-register also failed: " + e2.getMessage());
            }
        }
    }
}

File: xposed-pine/src/org/pixel/customparts/hooks/systemui/DozeTapDozeHook.kt
package org.pixel.customparts.hooks.systemui

import android.content.Context
import de.robv.android.xposed.XC_MethodHook
import de.robv.android.xposed.XposedBridge
import de.robv.android.xposed.XposedHelpers
import org.pixel.customparts.core.BaseHook

class DozeTapDozeHook : BaseHook() {
    override val hookId = "DozeTapDozeHook"
    override val priority = 70

    private var dozeTapReason = 9

    override fun isEnabled(context: Context?): Boolean = true

    override fun onInit(classLoader: ClassLoader) {
        resolveTapReason(classLoader)
        hookDozeTriggers(classLoader)
    }

    private fun resolveTapReason(classLoader: ClassLoader) {
        try {
            val dozeLogClass = XposedHelpers.findClass("com.android.systemui.doze.DozeLog", classLoader)
            dozeTapReason = XposedHelpers.getStaticIntField(dozeLogClass, "REASON_SENSOR_TAP")
        } catch (_: Throwable) {
            log("DozeTapDozeHook: Using default REASON_SENSOR_TAP = $dozeTapReason")
        }
    }

    private fun hookDozeTriggers(classLoader: ClassLoader) {
        try {
            val dozeTriggersClass = XposedHelpers.findClass(
                "com.android.systemui.doze.DozeTriggers",
                classLoader
            )

            XposedBridge.hookAllMethods(dozeTriggersClass, "onSensor", object : XC_MethodHook() {
                override fun beforeHookedMethod(param: MethodHookParam) {
                    try {
                        val pulseReason = param.args.firstOrNull { it is Int } as? Int ?: return
                        if (pulseReason != dozeTapReason) return

                        val dozeTriggers = param.thisObject
                        val context = XposedHelpers.getObjectField(dozeTriggers, "mContext") as? Context ?: return
                        if (!isSettingEnabled(context, DozeTapKeys.KEY_HOOK)) return

                        val floatArgs = param.args.filterIsInstance<Float>()
                        val screenX = floatArgs.getOrNull(0) ?: return
                        val screenY = floatArgs.getOrNull(1) ?: return

                        val timeout = getIntSetting(context, DozeTapKeys.KEY_TIMEOUT, DozeTapKeys.DEFAULT_TIMEOUT)
                        val consumed = DozeTapManager.processTap(
                            context,
                            screenX,
                            screenY,
                            true,
                            timeout
                        ) {
                            reregisterTapSensor(dozeTriggers)
                        }

                        if (consumed) {
                            param.result = null
                        }
                    } catch (t: Throwable) {
                        log("DozeTapDozeHook: Error in onSensor: ${t.message}")
                    }
                }
            })

            log("DozeTapDozeHook: Hook applied successfully")
        } catch (e: Throwable) {
            log("DozeTapDozeHook: Failed to apply hook: ${e.message}")
        }
    }

    private fun reregisterTapSensor(dozeTriggers: Any) {
        val dozeSensors = XposedHelpers.getObjectField(dozeTriggers, "mDozeSensors") ?: return
        val triggerSensors = XposedHelpers.getObjectField(dozeSensors, "mTriggerSensors") ?: return

        if (triggerSensors is Array<*>) {
            for (sensor in triggerSensors) {
                if (sensor == null) continue
                checkAndResetSensor(sensor)
            }
        } else if (triggerSensors is Collection<*>) {
            for (sensor in triggerSensors) {
                if (sensor == null) continue
                checkAndResetSensor(sensor)
            }
        }
    }

    private fun checkAndResetSensor(sensor: Any) {
        try {
            val reason = XposedHelpers.getIntField(sensor, "mPulseReason")
            if (reason == dozeTapReason) {
                val method = sensor.javaClass.getDeclaredMethod("setListening", Boolean::class.javaPrimitiveType)
                method.isAccessible = true
                method.invoke(sensor, false)
                method.invoke(sensor, true)
                log("DozeTapDozeHook: Sensor re-registered OK")
            }
        } catch (e: Throwable) {
            log("DozeTapDozeHook: checkAndResetSensor failed: ${e.message}")
            try {
                XposedHelpers.setBooleanField(sensor, "mRequested", false)
                XposedHelpers.callMethod(sensor, "updateListening")
                XposedHelpers.setBooleanField(sensor, "mRequested", true)
                XposedHelpers.callMethod(sensor, "updateListening")
                log("DozeTapDozeHook: Sensor re-registered via fallback OK")
            } catch (e2: Throwable) {
                log("DozeTapDozeHook: Fallback re-register also failed: ${e2.message}")
            }
        }
    }
}
